### PR TITLE
[loki] The new module

### PR DESCRIPTION
**Pull request by trublast**
## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: loki
type: feature
summary: The new module. Based on the Grafana Loki project.
```

### CODE REVIEW

File: modules/041-linstor/templates/linstor-scheduler-admission/webhook.yaml [hunk 1]
```diff
@@ -45,6 +45,7 @@ webhooks:
       - d8-descheduler
       - d8-ingress-nginx
       - d8-log-shipper
+      - d8-loki
```
Comment (kvaps) [inline modules/041-linstor/templates/linstor-scheduler-admission/webhook.yaml:48]: loki can use persistent volumes, please add it to exclusions of linstor-scheduler-admission webhook generator

https://github.com/deckhouse/deckhouse/blob/16d6f2c524e913e88fb356c0253c9dde4a3f9517/tools/linstor_scheduler_webhook/main.go#L32-L37

**Review by kvaps — CHANGES_REQUESTED** [review]

File: modules/462-loki/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+title: "The loki module"
+---
+
+[Loki](https://grafana.com/oss/loki/) - module for storing and viewing logs.
```
Comment (z9r5) [inline modules/462-loki/docs/README.md:5]:
```suggestion
[Loki](https://grafana.com/oss/loki/) - module for storing and viewing logs.
```
Does the module allow you to view logs?  Maybe  - "The module for aggregating logs"?

File: modules/462-loki/docs/README_RU.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+title: "Модуль loki"
+---
+
+[Loki](https://grafana.com/oss/loki/) - модуль для хранения и просмотра логов.
```
Comment (z9r5) [inline modules/462-loki/docs/README_RU.md:5]:
```suggestion
Модуль предназначен для организации хранилища логов.

Модуль использует проект [Grafana Loki](https://grafana.com/oss/loki/).

Модуль разворачивает хранилище логов на базе Grafana Loki, при необходимости настраивает модуль log-shipper на использование loki и добавляет в Grafana соответствующий datasource. 
```

File: modules/462-loki/openapi/config-values.yaml [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+properties:
+  effectiveStorageClass:
+    oneOf:
+      - type: string
+      - type: boolean
+    x-doc-default: 'false'
+  diskSizeGigabytes:
+    type: integer
+    description: |
+      Disk size for log store. PVC size if `effectiveStorageClass` set.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:11]:
```suggestion
      Disk size for storing logs or the size of the PVC when specifying the [storageClass](#parameters-storageclass) parameters.
```
It seems that the effectiveStorageClass parameter have te be named as storageClass.

File: modules/462-loki/openapi/config-values.yaml [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    description: |
+      How many hours logs will be kept before they are deleted.
+    x-doc-default: 168
+    default: 168
+  storeSystemLogs:
+    type: boolean
+    default: true
+    x-doc-default: 'true'
+    description: |
+      Store logs from d8-* namespaces into loki. [log-shipper](../460-log-shipper) module must be enabled.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:25]:
```suggestion
      Store logs from `d8-*` namespaces into loki.

      The [log-shipper](../460-log-shipper) module must be enabled.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    additionalProperties:
+      type: string
+    description: |
+      The same as in the Pods' `spec.nodeSelector` parameter in Kubernetes.
+
+      If the parameter is omitted or `false`, `nodeSelector` will be determined [automatically](https://deckhouse.io/documentation/v1/#advanced-scheduling).
+  tolerations:
+    type: array
+    description: |
+      The same as in the Pods' `spec.tolerations` parameter in Kubernetes;
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:37]:
```suggestion
      The same as in the Pods' `spec.tolerations` parameter in Kubernetes.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+          format: int64
+        value:
+          type: string
+  resourcesRequests:
+    required: ["mode"]
+    type: object
+    description: |
+      Max amounts of CPU and memory resources that the pod can request when selecting a node.
+
+      If the `vertical-pod-autoscaler` module is disabled, then these values become the default ones.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:60]:
```suggestion
      Max amounts of CPU and memory resources that the Pod can request when selecting a node.

      If the [vertical-pod-autoscaler](../302-vertical-pod-autoscaler/) module is disabled, then these values become the default ones.
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 8]
```diff
@@ -0,0 +1,70 @@
+type: object
+properties:
+  diskSizeGigabytes:
+    description: |
+      Размер диска для хранения логов. Размер PVC при указании `effectiveStorageClass`
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:5]:
```suggestion
      Размер диска для хранения логов или размер PVC при указании параметра [storageClass](#parameters-storageclass).
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  retentionPeriodHours:
+    description: |
+      Сколько часов логи будут храниться прежде чем будут удалены.
+  storeSystemLogs:
+    type: boolean
+    default: true
+    x-doc-default: 'true'
+    description: |
+      Сохранять логи из d8-* неймспейсов в loki. Должен быть включен модуль [log-shipper](../460-log-shipper).
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:14]:
```suggestion
      Сохранять логи из пространства имен `d8-*` в loki.

      Для работы необходим включенный модуль [log-shipper](../460-log-shipper).
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+  tolerations:
+    description: |
+      Структура, аналогичная  `spec.tolerations` в Kubernetes Pod.
+
+      Если ничего не указано или указано `false` — будет [использоваться автоматика](https://deckhouse.ru/documentation/v1/#выделение-узлов-под-определенный-вид-нагрузки).
+  resourcesRequests:
+    description: |
+      Настройки максимальных значений CPU и memory, которые может запросить Pod при выборе узла.
+
+      Если модуль `vertical-pod-autoscaler` выключен, максимальные значения становятся значениями по умолчанию.
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:29]:
```suggestion
      Если модуль [vertical-pod-autoscaler](../302-vertical-pod-autoscaler/) выключен, то максимальные значения становятся значениями по умолчанию.
```

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/stage'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'deploy/web/test'...

File: modules/462-loki/openapi/config-values.yaml [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        enum: [false]
+    x-examples: [false, "default"]
+    description: |
+      The name of the StorageClass to use.
+
+      If omitted, the StorageClass of the existing Loki PVC is used. If there is no PVC yet, either `global.StorageClass` or `global.discovery.defaultStorageClass` is used, and if those are undefined, the emptyDir volume is used to store the data.
+
+      `storageClass: false` — forces the `emptyDir` usage. You will need to delete the old PVC and restart the Pod manually.
+
+      **CAUTION!** Setting this value to one that differs from the current one (in the existing PVC) will result in Loki volume reprovisioning and data loss.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:16]:
```suggestion
      The name of the StorageClass to use.

      If omitted, the StorageClass of the existing Loki PVC is used. If there is no PVC yet, either the global [StorageClass](../../deckhouse-configure-global.html#parameters-storageclass) or `global.discovery.defaultStorageClass` is used, and if those are undefined, the emptyDir volume is used to store the data.

      `storageClass: false` — forces the `emptyDir` usage. You will need to delete the old PVC and restart the Pod manually.

      **CAUTION!** Setting this value to one that differs from the current one (in the existing PVC) will result in Loki volume reprovisioning and data loss.
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 12]
```diff
@@ -0,0 +1,81 @@
+type: object
+properties:
+  storageClass:
+    description: |
+      Имя storageClass'а, который использовать.
+
+      Если не указано — используется StorageClass существующей PVC Loki, а если PVC пока нет — используется или `global.storageClass`, или `global.discovery.defaultStorageClass`, а если и их нет — данные сохраняются в `emptyDir`.
+
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:8]:
```suggestion
      Если не указано — используется StorageClass существующей PVC Loki, а если PVC пока нет — используется или  [глобальный StorageClass](../../deckhouse-configure-global.html#parameters-storageclass), или `global.discovery.defaultStorageClass`, а если и их нет — данные сохраняются в `emptyDir`.

```

**Review by z9r5 — DISMISSED** [review]

File: modules/462-loki/openapi/config-values.yaml [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+  diskSizeGigabytes:
+    type: integer
+    description: |
+      Disk size for storing logs or the size of the PVC when specifying the [storageClass](#parameters-storageclass) parameters.
+
+      You should manually specify the required disk size for storing logs for the period specified in the `retentionPeriodHours` parameter.
+      Loki does not have the ability to automatically delete old data when the disk is full.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:23]:
```suggestion
      Loki cannot automatically delete old data when the disk is full.
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+      `false` — принудительное использование `emptyDir`. Удалить старый PVC и рестартануть Pod придется руками
+
+      **ОСТОРОЖНО!** При указании этой опции в значение, отличное от текущего (из существующей PVC), диск Loki будет перезаказан, а все данные удалены.
+  diskSizeGigabytes:
+    description: |
+      Размер диска для хранения логов или размер PVC при указании параметра [storageClass](#parameters-storageclass).
+
+      Вы должны самостоятельно указать необходимый размер диска для хранения логов за период, указанный в параметре `retentionPeriodHours`.
+      Loki не имеет возможности производить удаление старых данных автоматически при заполнении диска.
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:17]:
```suggestion
      Loki не может удалять старые данные автоматически при заполнении диска.
```

File: modules/462-loki/openapi/config-values.yaml (hunk 13 above)
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:23]:
```suggestion
      Loki cannot automatically delete old data when the disk is full.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+
+      The occupied disk space is automatically monitored by [extended-monitoring](../340-extended-monitoring/configuration.html#namespaced-kubernetes-objects) module.
+
+      **CAUTION!**
+      If the Loki data disk becomes full, the application will stop working. You should choose the necessary balance between the values of
+      `retentionPeriodHours` and `diskSizeGigabytes`.
```
Comment (z9r5) [inline modules/462-loki/openapi/config-values.yaml:29]:
```suggestion
      **CAUTION!** If the Loki data disk becomes full, the application will stop working. You should choose the necessary balance between the values of `retentionPeriodHours` and `diskSizeGigabytes`.
```

File: modules/462-loki/openapi/doc-ru-config-values.yaml [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+
+      Занятый объем автоматически мониторится с помощью модуля [extended-monitoring](../340-extended-monitoring/configuration.html#namespaced-kubernetes-objects)
+
+      **ОСТОРОЖНО!**
+      Если диск с данными Loki переполнится - приложение перестанет работать. Вам следует выбрать необходимый баланс между значениями параметров
+      `retentionPeriodHours` и `diskSizeGigabytes`.
```
Comment (z9r5) [inline modules/462-loki/openapi/doc-ru-config-values.yaml:23]:
```suggestion
      **ОСТОРОЖНО!**  Если диск с данными Loki переполнится, то приложение перестанет работать. Вам следует выбрать необходимый баланс между значениями параметров `retentionPeriodHours` и `diskSizeGigabytes`.
```

**Review by RomanenkoDenys — COMMENTED** [review]
LGTM

**Review by RomanenkoDenys — APPROVED** [review]

File: modules/462-loki/openapi/config-values.yaml [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    description: |
+      How many hours logs will be kept before they are deleted.
+    x-doc-default: 168
+    default: 168
+  storeSystemLogs:
+    type: boolean
+    default: true
+    x-doc-default: 'true'
+    description: |
+      Store logs from `d8-*` namespaces into loki.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:41]:
```suggestion
      Save logs from the `d8-*` namespaces to loki.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+
+      The [log-shipper](../460-log-shipper) module must be enabled.
+  nodeSelector:
+    type: object
+    additionalProperties:
+      type: string
+    description: |
+      The same as in the Pods' `spec.nodeSelector` parameter in Kubernetes.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:49]:
```suggestion
      The same as the Pods' `spec.nodeSelector` parameter in Kubernetes.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+
+      If the parameter is omitted or `false`, `nodeSelector` will be determined [automatically](https://deckhouse.io/documentation/v1/#advanced-scheduling).
+  tolerations:
+    type: array
+    description: |
+      The same as in the Pods' `spec.tolerations` parameter in Kubernetes.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:55]:
```suggestion
      The same as the Pods' `spec.tolerations` parameter in Kubernetes.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+        tolerationSeconds:
+          type: integer
+          format: int64
+        value:
+          type: string
+  resourcesRequests:
+    required: ["mode"]
+    type: object
+    description: |
+      Max amounts of CPU and memory resources that the Pod can request when selecting a node.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:76]:
```suggestion
      The max amounts of CPU and memory resources the Pod can request when being scheduled on a node.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+              CPU-related parameters.
+            default: {}
+            properties:
+              max:
+                oneOf:
+                  - type: string
+                    pattern: "^[0-9]+m?$"
+                  - type: number
+                description: |
+                  Maximum allowed CPU requests.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:147]:
```suggestion
                  The max amount of CPU requests the VPA can set.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+                default: "500m"
+              min:
+                oneOf:
+                  - type: string
+                    pattern: "^[0-9]+m?$"
+                  - type: number
+                description: |
+                  Minimum allowed CPU requests.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:155]:
```suggestion
                  The min amount of CPU requests the VPA can set.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+              The amount of memory requested.
+            default: {}
+            properties:
+              max:
+                oneOf:
+                  - type: string
+                    pattern: '^[0-9]+(\.[0-9]+)?(E|P|T|G|M|k|Ei|Pi|Ti|Gi|Mi|Ki)?$'
+                  - type: number
+                description: |
+                  Maximum allowed memory requests.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:169]:
```suggestion
                  The max amount of memory requests the VPA can set.
```

File: modules/462-loki/openapi/config-values.yaml [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+                default: "2048Mi"
+              min:
+                oneOf:
+                  - type: string
+                    pattern: '^[0-9]+(\.[0-9]+)?(E|P|T|G|M|k|Ei|Pi|Ti|Gi|Mi|Ki)?$'
+                  - type: number
+                description: |
+                  Minimum allowed memory requests.
```
Comment (kirkonru) [inline modules/462-loki/openapi/config-values.yaml:177]:
```suggestion
                  The min amount of memory requests the VPA can set.
```

File: modules/462-loki/openapi/config-values.yaml (hunk 21 above)
Comment (kirkonru) [reply modules/462-loki/openapi/config-values.yaml:147]: I'm not sure if I get this right, but it's still worth a try.

**Review by kirkonru — APPROVED** [review]

**Review by z9r5 — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/aws'...

File: modules/462-loki/docs/README_RU.md [hunk 25]
```diff
@@ -0,0 +1,9 @@
+---
+title: "Модуль loki"
+---
+
+Модуль предназначен для организации хранилища логов.
+
+Модуль использует проект [Grafana Loki](https://grafana.com/oss/loki/).
+
+Модуль разворачивает хранилище логов на базе Grafana Loki, при необходимости настраивает модуль log-shipper на использование loki и добавляет в Grafana соответствующий datasource.
```
Comment (z9r5) [inline modules/462-loki/docs/README_RU.md:9]:
```suggestion
Модуль разворачивает хранилище логов на базе Grafana Loki, при необходимости настраивает модуль [log-shipper](../460-log-shipper/) на использование loki и добавляет в Grafana соответствующий datasource.
```

File: modules/462-loki/docs/README.md [hunk 26]
```diff
@@ -0,0 +1,9 @@
+---
+title: "The loki module"
+---
+
+The module implements log storage.
+
+The module uses the [Grafana Loki](https://grafana.com/oss/loki/) project.
+
+The module deploys log storage based on Grafana Loki, configures the [log-shipper](../460-log-shipper) module to use loki if necessary, and adds the corresponding data source to Grafana.
```
Comment (z9r5) [inline modules/462-loki/docs/README.md:9]:
```suggestion
The module deploys log storage based on Grafana Loki, configures the [log-shipper](../460-log-shipper/) module to use loki if necessary, and adds the corresponding data source to Grafana.
```

File: modules/462-loki/templates/statefulset.yaml [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+        checksum/config: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
+    spec:
+      imagePullSecrets:
+      - name: deckhouse-registry
+      terminationGracePeriodSeconds: 4800
+      {{- include "helm_lib_node_selector" (tuple . "system") | nindent 6 }}
+      {{- include "helm_lib_tolerations" (tuple . "system") | nindent 6 }}
+      {{- include "helm_lib_priority_class" (tuple . "cluster-low") | nindent 6 }}
+      {{- include "helm_lib_module_pod_security_context_run_as_user_nobody_with_writable_fs" . | nindent 6 }}
+      containers:
```
[the remark this answers is not in the document]
Comment (nabokihms) [reply modules/462-loki/templates/statefulset.yaml:57]: We can add it in the future.

[the remark this answers is not in the document]
Comment (nabokihms) [reply modules/462-loki/templates/statefulset.yaml:39]: https://github.com/deckhouse/deckhouse/issues/4315
https://github.com/deckhouse/deckhouse/issues/4314

**Pull request merged**